### PR TITLE
Add API that allows the emitter to indicate presence of localloc

### PR DESCRIPTION
**Pull request by tmat**
When encoding method body header the encoder needs to decide whether to emit tiny or fat header. A small method that has no locals but contains `localloc` instruction and has InitLocals set to true should not be encoded with tiny header since tiny header implies InitLocals is false. The presence of `localloc` instruction needs to be indicated by the caller of the method body encoder similarly to max stack and other info. Hence we need to add an overload that takes an extra bool parameter.

Fixes https://github.com/dotnet/corefx/issues/26910
Implements https://github.com/dotnet/corefx/issues/27611

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera @VSadov PTAL

**Comment by nguerrera** [comment]
LGTM. This is technically new API needing review. @terrajobst can we fast track it?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs [hunk 1]
```diff
@@ -114,11 +138,35 @@ internal MethodBody(int bodyOffset, Blob instructions, ExceptionRegionEncoder ex
...
+        /// <param name="localVariablesSignature">Local variables signature handle.</param>
+        /// <param name="attributes">Attributes.</param>
+        /// <param name="hasDynamicStackAllocation">True if the method allocates from dynamic local memory pool (the IL contains <c>localloc</c> instruction).
+        /// </param>
+        /// <returns>The offset of the encoded body within the method body stream.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="instructionEncoder"/> has default value.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="maxStack"/> is out of range [0, <see cref="ushort.MaxValue"/>].</exception>
+        /// <exception cref="InvalidOperationException">
+        /// A label targeted by a branch in the instruction stream has not been marked,
+        /// or the distance between a branch instruction and the target label is doesn't fit the size of the instruction operand.
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: Typo: "is doesn't"

**Comment by VSadov** [comment]
Cc:@t-camaia, @agocke, @jaredpar

**Comment by ahsonkhan** [comment]
API changes here. What's the plan here?
cc @terrajobst, @karelz, @danmosemsft

**Comment by danmosemsft** [comment]
@ahsonkhan when API is approved send shiproom brief note with reasoning for the churn.

**Comment by karelz** [comment]
I don't see ref changes ... should they be there as well?

**Comment by agocke** [comment]
cc @t-camaia

**Comment by tmat** [comment]
@karelz Created API proposal issue: https://github.com/dotnet/corefx/issues/27611

**Comment by tmat** [comment]
test Tizen armel Debug Build please

**Comment by VSadov** [comment]
@agocke - we can also always set hasDynamicStackAllocation and just twiddle the LocalsInit in the attributes. - the result is the same.
Or track the presence of stackalloc. We would have that information anyways, just would need to plumb it through.

Yes the API change is additive to not be breaking.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs (hunk 1 above)
Comment (tmat) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: Fixed.

**Comment by tmat** [comment]
> I don't see ref changes ... should they be there as well?

Are they not auto-generated? I run `build-managed.cmd`, it succeeded and don't see any other files changed in the repo.

**Comment by ahsonkhan** [comment]
> Are they not auto-generated? I run `build-managed.cmd`, it succeeded and don't see any other files changed in the repo.

Try running `msbuild /t:GenerateReferenceSource /p:TargetGroup=netstandard /p:ConfigurationGroup=Release` (from src directory of your project - in this case https://github.com/tmat/corefx/tree/27c4627148084a29bb6839e3a39a7d9d05767de6/src/System.Reflection.Metadata/src)

This is after building the repo at the root.

**Comment by tmat** [comment]
@ahsonkhan Thanks. Added another commit with regenerated ref file.

This process is very much broken. My previous PR also added new APIs but I didn't know the ref files need to be manually regenerated. Yet everything passed and it got merged!
The regenerated ref file also contains some other changes that have likely been made previously without regenerating the ref file: https://github.com/dotnet/corefx/pull/27589/commits/df10420a98c2c1a8db73021b9bfd2c75182d02d8

I wonder how many bugs are there currently in CoreFX due to this ref file generation.

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 2]
```diff
@@ -99,9 +99,7 @@ namespace System.Reflection.Metadata
         public System.Reflection.Metadata.StringHandle Name { get { throw null; } }
         public System.Reflection.Metadata.BlobHandle PublicKey { get { throw null; } }
         public System.Version Version { get { throw null; } }
-#if !NETSTANDARD11
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: You need to sanitize the changes so only include the necessary changes you want.

Keep the ifdefs.

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 3]
```diff
@@ -2039,7 +2036,7 @@ public enum PrimitiveTypeCode : byte
     }
     public readonly partial struct PropertyAccessors
     {
-        private readonly int _dummy;
+        private readonly object _dummy;
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: @weshaggard, is this change OK and expected (the _dummy fields going from int to object)?

Comment (weshaggard) [reply src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: Some of those definitely look suspicious. object is the fallback if we cannot determine all the fields in the struct.

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 4]
```diff
@@ -3135,6 +3138,7 @@ public enum Machine : ushort
         AM33 = (ushort)467,
         Amd64 = (ushort)34404,
         Arm = (ushort)448,
+        Arm64 = (ushort)43620,
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: @davidwrighton, this was added here - https://github.com/dotnet/corefx/pull/26243

This is expected, correct?

**Comment by weshaggard** [comment]
>This process is very much broken. My previous PR also added new APIs but I didn't know the ref files need to be manually regenerated

The process isn't the greatest but it is the best we have. We cannot automatically generate these easily because we have a different surface area based on configuration in a lot of libraries there isn't one definitively implementation we can generate them from. 

>I wonder how many bugs are there currently in CoreFX due to this ref file generation.

There may be a few but I don't expect there is a lot because most folks try to test or use their new APIs and they do it via the ref assembly and if the APIs aren't present then they cannot use them.

**Comment by ahsonkhan** [comment]
> This process is very much broken. My previous PR also added new APIs but I didn't know the ref files need to be manually regenerated. Yet everything passed and it got merged!

If you add tests that exercise the newly added APIs, the build won't pass without the APIs being added to the ref.

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs (hunk 3 above)
Comment (ahsonkhan) [reply src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: Can you provide some rudimentary guidelines that we can follow to vet these changes ourselves?

**Comment by ahsonkhan** [comment]
> There may be a few but I don't expect there is a lot because most folks try to test or use their new APIs and they do it via the ref assembly and if the APIs aren't present then they cannot use them.

Precisely.

https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/tests/System.Reflection.Metadata.Tests.csproj#L144
```csproj
<!-- Some internal types are needed, so we reference the implementation assembly, rather than the reference assembly. -->
```

**Comment by tmat** [comment]
> There may be a few but I don't expect there is a lot because most folks try to test or use their new APIs and they do it via the ref assembly and if the APIs aren't present then they cannot use them.

The unit tests pass :(

**Comment by ahsonkhan** [comment]
> The unit tests pass :(

I wonder if we can use InternalsVisibleTo instead of referencing the implementation assembly. Would we then be able to reference the ref and circumvent this issue?

**Comment by tmat** [comment]
> The process isn't the greatest but it is the best we have. 

This is not an acceptable answer. 
@Petermarcu

**Comment by danmosemsft** [comment]
Seems the problem is not that one must manually update the ref, but that by some unknown means the unit tests passed without the ref update. @weshaggard ?

**Comment by nguerrera** [comment]
@danmosemsft The means has been identified: the tests reference the implementation assembly as they require access to internals.

**Comment by nguerrera** [comment]
> I wonder if we can use InternalsVisibleTo instead of referencing the implementation assembly. Would we then be able to reference the ref and circumvent this issue?

We are using InternalsVisibleTo. Do you mean put the internals in the reference assembly? Do other assemblies do that? If this closes the test gap, then yes, we should at least do that.

**Comment by tmat** [comment]
Filed an issue to fix the process: https://github.com/dotnet/corefx/issues/27639

**Comment by tmat** [comment]
Unit tests won't validate that the ref assembly is 100% correct. E.g. ref assembly contains private dummy fields. No unit test will validate those.

**Comment by tmat** [comment]
I have somewhat fixed up the ref assembly. Please validate it's correct.

**Comment by ahsonkhan** [comment]
> I have somewhat fixed up the ref assembly. Please validate it's correct.

The only thing would be to keep the `#if !NETSTANDARD11` that were removed. See https://github.com/dotnet/corefx/pull/27589/files#r171741403

**Comment by danmosemsft** [comment]
What about two test libraries for the same implementation library, one with  IVT and the other going only through public API would that help?

**Comment by tmat** [comment]
@danmosemsft. That makes it more complicated, not less. I don't want any more complexity in this already over complicated system. 

This would help:
1) Generate one file per configuration and target framework 
2) Generate the file to intermediate directory and do not check it in
3) Use C# compiler ref assembly feature

Perhaps these are not applicable on all projects in CoreFX that need some magic like System.Runtime but it will work at least for SCI and SRM (and maybe for others).

**Comment by tmat** [comment]
Let's move the ref assembly discussion to https://github.com/dotnet/corefx/issues/27639

**Comment by tmat** [comment]
test OSX x64 Debug Build please

**Comment by tmat** [comment]
test UWP CoreCLR x64 Debug Build please

**Comment by tmat** [comment]
test OSX x64 Debug Build please

**Comment by tmat** [comment]
test UWP CoreCLR x64 Debug Build please

**Comment by ahsonkhan** [comment]
The OSX failure is a known issue - https://github.com/dotnet/core-eng/issues/2808

**Comment by weshaggard** [comment]
While working on https://github.com/dotnet/corefx/issues/27639 I regenerated the ref for System.Reflection.Metadata and also noticed the dummy fields being changed from int to object and I do think that is a correct update because those structs have ImmutableArray references which contain array references. As part of addressing https://github.com/dotnet/corefx/issues/27639 I will regenerate the ref to fix that.

**Pull request merged**